### DISCUSSION TITLE
VSCode extension / Language Server

### DISCUSSION THREAD

**Discussion by michalczaplinski** [WordPress/gutenberg#53036] [Interactivity API]
I would like to explore the idea of a VSCode extension (based on [LSP](https://microsoft.github.io/language-server-protocol/) so it can be used in other IDEs too) that could detect the uses of `wpx` in your project and give you syntax highlighting, autocompletions, type information, etc. in your PHP files. 

There is some prior work for similar frameworks:
- Vue - https://github.com/vuejs/vetur
- Svelte - https://github.com/sveltejs/language-tools
- Alpine - https://github.com/AdrianWilczynski/AlpineIntelliSense (this one is not official and not well-maintained) 

@DAreRodz based on your work in https://github.com/woocommerce/woocommerce-blocks/pull/8221 I'd be curious to hear about the DX of working with the JS object created with `wpx` and using it in PHP in a separate file. Is it a hassle to have to use use "stringly" values like `"actions.filters.updateActiveHandle"`? I assume that there are no IDE hints nor autocompletion, right? Anything else that seemed like a hassle compared to just working with (P)React?

**Comment by DAreRodz** [comment]
Good questions. 🤔

I have yet to experience any hassle so far. However, I had to admit that it feels kind of different to React or Preact DX, mainly because the state (and the logic to interact with it) lives in one file (`view.js`), and the template consuming the state and event handlers in another (either in `render.php` or in `save.js`). Well, not so different if you're used to state manager libraries.

Apart from those string "selectors", when you want to have some sort of local state (like using `useState` inside a component), you would need `wp-context` to achieve something similar―and that context needs to be defined in the template as a stringified JSON. This case is not a hard problem to solve, though, because you can define an object (both in JS or PHP, whatever you're using for your template) and stringify it to JSON once you're going to use it in the JSX/HTML code.

In any case, if you want to use those selectors to access either the state or the current context, it would be useful to have autocompletion hints. 😄

**Reply by michalczaplinski** [reply]
ok, I see! Thanks @DAreRodz that's useful information. I think it's definitely too early to work on this but I'd like to keep thinking about it 🙂 I would love if the DX of using Directives were on par with React there.

**Comment by michalczaplinski** [comment]
@DAreRodz, @ockham  and I have hacked on [a VSCode extension](https://github.com/michalczaplinski/interactivity-vscode) that provides autocompletion for the `state`, `actions`, `selectors`, and `effects` present in the store on the fronted.

It is currently just a prototype and neither of us is actively working on it at the moment, but I'd like to push it forward in the coming months.
